File: lib/widgets/fields.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../theme/theme.dart';

class QTextField extends StatelessWidget {
  const QTextField({
    super.key,
    required this.controller,
    required this.label,
    this.hint,
    this.obscure = false,
    this.maxLines = 1,
    this.keyboardType,
    this.inputFormatters,
    this.autofocus = false,
    this.onSubmitted,
  });

  final TextEditingController controller;
  final String label;
  final String? hint;
  final bool obscure;
  final int maxLines;
  final TextInputType? keyboardType;
  final List<TextInputFormatter>? inputFormatters;
  final bool autofocus;
  final ValueChanged<String>? onSubmitted;

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    OutlineInputBorder border(Color color, [double width = 1]) =>
        OutlineInputBorder(
          borderRadius: BorderRadius.circular(12),
          borderSide: BorderSide(color: color, width: width),
        );

    return TextField(
      controller: controller,
      obscureText: obscure,
      maxLines: obscure ? 1 : maxLines,
      keyboardType: keyboardType,
      inputFormatters: inputFormatters,
      autofocus: autofocus,
      onSubmitted: onSubmitted,
      style: TextStyle(color: colors.textPrimary, fontSize: 15),
      decoration: InputDecoration(
        labelText: label,
        hintText: hint,
        hintStyle: TextStyle(color: colors.textMuted, fontSize: 14),
        labelStyle: TextStyle(color: colors.textMuted, fontSize: 14),
        filled: true,
        fillColor: colors.card,
        isDense: true,
        contentPadding: const EdgeInsets.symmetric(
          horizontal: 14,
          vertical: 14,
        ),
        border: border(colors.divider),
        enabledBorder: border(colors.divider),
        focusedBorder: border(colors.accent, 1.5),
      ),
    );
  }
}

class QFormLabel extends StatelessWidget {
  const QFormLabel(this.text, {super.key});

  final String text;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(2, 18, 2, 8),
      child: Text(
        text.toUpperCase(),
        style: TextStyle(
          color: context.appColors.textMuted,
          fontSize: 11,
          fontWeight: FontWeight.w700,
          letterSpacing: 0.6,
        ),
      ),
    );
  }
}

class QEmptyView extends StatelessWidget {
  const QEmptyView({
    super.key,
    required this.icon,
    required this.title,
    required this.message,
    this.action,
  });

  final IconData icon;
  final String title;
  final String message;
  final Widget? action;

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    return Center(
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 40),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Icon(icon, color: colors.textMuted, size: 44),
            const SizedBox(height: 16),
            Text(
              title,
              textAlign: TextAlign.center,
              style: TextStyle(
                color: colors.textSecondary,
                fontSize: 16,
                fontWeight: FontWeight.w600,
              ),
            ),
            const SizedBox(height: 6),
            Text(
              message,
              textAlign: TextAlign.center,
              style: TextStyle(
                color: colors.textMuted,
                fontSize: 13,
                height: 1.35,
              ),
            ),
            if (action != null) ...[const SizedBox(height: 18), action!],
          ],
        ),
      ),
    );
  }
}

Future<bool> confirmDestructive(
  BuildContext context, {
  required String title,
  required String message,
  String actionLabel = 'Delete',
}) async {
  final colors = context.appColors;
  final result = await showDialog<bool>(
    context: context,
    barrierColor: colors.dialogBarrier,
    builder: (dialogContext) => AlertDialog(
      backgroundColor: colors.dialogBackground,
      title: Text(title, style: TextStyle(color: colors.dialogText)),
      content: Text(message, style: TextStyle(color: colors.dialogMuted)),
      actions: [
        TextButton(
          onPressed: () => Navigator.of(dialogContext).pop(false),
          child: const Text('Cancel'),
        ),
        TextButton(
          onPressed: () => Navigator.of(dialogContext).pop(true),
          style: TextButton.styleFrom(foregroundColor: colors.danger),
          child: Text(actionLabel),
        ),
      ],
    ),
  );
  return result ?? false;
}

class PickOption<T> {
  const PickOption(
    this.value,
    this.label, {
    this.subtitle,
    this.icon,
    this.isAction = false,
  });

  final T value;
  final String label;
  final String? subtitle;
  final IconData? icon;

  final bool isAction;
}

Future<PickOption<T>?> pickFromList<T>(
  BuildContext context, {
  required String title,
  required List<PickOption<T>> options,
  required T current,
}) {
  final colors = context.appColors;
  return showDialog<PickOption<T>>(
    context: context,
    barrierColor: colors.dialogBarrier,
    builder: (dialogContext) => AlertDialog(
      backgroundColor: colors.dialogBackground,
      title: Text(title, style: TextStyle(color: colors.dialogText)),
      contentPadding: const EdgeInsets.fromLTRB(0, 12, 0, 8),
      content: SizedBox(
        width: 340,
        child: ConstrainedBox(
          constraints: BoxConstraints(
            maxHeight: MediaQuery.of(dialogContext).size.height * 0.5,
          ),
          child: ListView.builder(
            shrinkWrap: true,
            itemCount: options.length,
            itemBuilder: (context, index) {
              final option = options[index];
              final selected = !option.isAction && option.value == current;
              final tint = option.isAction ? colors.accent : colors.dialogText;
              return ListTile(
                dense: true,
                leading: option.icon == null
                    ? null
                    : Icon(
                        option.icon,
                        size: 20,
                        color: option.isAction
                            ? colors.accent
                            : colors.textMuted,
                      ),
                title: Text(
                  option.label,
                  style: TextStyle(
                    color: tint,
                    fontSize: 14,
                    fontWeight: option.isAction
                        ? FontWeight.w600
                        : FontWeight.w400,
                  ),
                ),
                subtitle: option.subtitle == null
                    ? null
                    : Text(
                        option.subtitle!,
                        style: TextStyle(
                          color: colors.dialogMuted,
                          fontSize: 12,
                        ),
                      ),
                trailing: selected
                    ? Icon(Icons.check, size: 18, color: colors.accent)
                    : null,
                onTap: () => Navigator.of(dialogContext).pop(option),
              );
            },
          ),
        ),
      ),
      actions: [
        TextButton(
          onPressed: () => Navigator.of(dialogContext).pop(),
          child: const Text('Cancel'),
        ),
      ],
    ),
  );
}

Future<String?> promptForText(
  BuildContext context, {
  required String title,
  required String label,
  String initial = '',
  bool obscure = false,
  String actionLabel = 'Save',
}) {
  return showDialog<String>(
    context: context,
    barrierColor: context.appColors.dialogBarrier,
    builder: (_) => _TextPrompt(
      title: title,
      label: label,
      initial: initial,
      obscure: obscure,
      actionLabel: actionLabel,
    ),
  );
}

/// The controller lives on the [State]: `showDialog`'s future completes on
/// pop(), but the dialog keeps rebuilding through its dismissal animation, so
/// disposing it in the caller uses it after disposal.
class _TextPrompt extends StatefulWidget {
  const _TextPrompt({
    required this.title,
    required this.label,
    required this.initial,
    required this.obscure,
    required this.actionLabel,
  });

  final String title;
  final String label;
  final String initial;
  final bool obscure;
  final String actionLabel;

  @override
  State<_TextPrompt> createState() => _TextPromptState();
}

class _TextPromptState extends State<_TextPrompt> {
  late final TextEditingController _controller = TextEditingController(
    text: widget.initial,
  );

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  void _submit() => Navigator.of(context).pop(_controller.text.trim());

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    return AlertDialog(
      backgroundColor: colors.dialogBackground,
      title: Text(widget.title, style: TextStyle(color: colors.dialogText)),
      content: QTextField(
        controller: _controller,
        label: widget.label,
        obscure: widget.obscure,
        autofocus: true,
        onSubmitted: (_) => _submit(),
      ),
      actions: [
        TextButton(
          onPressed: () => Navigator.of(context).pop(),
          child: const Text('Cancel'),
        ),
        TextButton(onPressed: _submit, child: Text(widget.actionLabel)),
      ],
    );
  }
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/widgets/Fields.kt
package com.example.ohmyssh.widgets

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.ListItemDefaults
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.ButtonDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.ohmyssh.theme.appColors
import com.example.ohmyssh.ui.Dialogs

@Composable
fun QTextField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    hint: String? = null,
    obscure: Boolean = false,
    maxLines: Int = 1,
    digitsOnly: Boolean = false,
    autofocus: Boolean = false,
    onSubmitted: ((String) -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    val colors = appColors
    val focusRequester = remember { FocusRequester() }

    OutlinedTextField(
        value = value,
        onValueChange = { raw ->
            onValueChange(if (digitsOnly) raw.filter { it.isDigit() } else raw)
        },
        label = { Text(label, fontSize = 14.sp) },
        placeholder = hint?.let { { Text(it, fontSize = 14.sp, color = colors.textMuted) } },
        visualTransformation = if (obscure) PasswordVisualTransformation() else VisualTransformation.None,
        singleLine = obscure || maxLines == 1,
        minLines = if (obscure) 1 else maxLines.coerceAtMost(3),
        maxLines = if (obscure) 1 else maxLines,
        keyboardOptions = KeyboardOptions(
            keyboardType = when {
                digitsOnly -> KeyboardType.Number
                obscure -> KeyboardType.Password
                else -> KeyboardType.Text
            },
        ),
        keyboardActions = KeyboardActions(onDone = { onSubmitted?.invoke(value) }),
        textStyle = TextStyle(color = colors.textPrimary, fontSize = 15.sp),
        shape = RoundedCornerShape(12.dp),
        colors = OutlinedTextFieldDefaults.colors(
            focusedContainerColor = colors.card,
            unfocusedContainerColor = colors.card,
            focusedBorderColor = colors.accent,
            unfocusedBorderColor = colors.divider,
            focusedLabelColor = colors.textMuted,
            unfocusedLabelColor = colors.textMuted,
            cursorColor = colors.accent,
        ),
        modifier = modifier.fillMaxWidth().focusRequester(focusRequester),
    )

    if (autofocus) {
        LaunchedEffect(Unit) { focusRequester.requestFocus() }
    }
}

@Composable
fun QFormLabel(text: String) {
    Text(
        text.uppercase(),
        style = TextStyle(
            color = appColors.textMuted,
            fontSize = 11.sp,
            fontWeight = FontWeight.W700,
            letterSpacing = 0.6.sp,
        ),
        modifier = Modifier.padding(start = 2.dp, top = 18.dp, end = 2.dp, bottom = 8.dp),
    )
}

@Composable
fun QEmptyView(
    icon: ImageVector,
    title: String,
    message: String,
    action: (@Composable () -> Unit)? = null,
) {
    val colors = appColors
    Box(Modifier.fillMaxWidth().heightIn(min = 240.dp), contentAlignment = Alignment.Center) {
        Column(
            Modifier.padding(horizontal = 40.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Icon(icon, contentDescription = null, tint = colors.textMuted, modifier = Modifier.size(44.dp))
            Spacer(Modifier.height(16.dp))
            Text(
                title,
                textAlign = TextAlign.Center,
                style = TextStyle(
                    color = colors.textSecondary,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.W600,
                ),
            )
            Spacer(Modifier.height(6.dp))
            Text(
                message,
                textAlign = TextAlign.Center,
                style = TextStyle(color = colors.textMuted, fontSize = 13.sp, lineHeight = 17.5.sp),
            )
            if (action != null) {
                Spacer(Modifier.height(18.dp))
                action()
            }
        }
    }
}

suspend fun confirmDestructive(
    title: String,
    message: String,
    actionLabel: String = "Delete",
): Boolean = Dialogs.show<Boolean> { dismiss ->
    val colors = appColors
    AlertDialog(
        onDismissRequest = { dismiss(false) },
        containerColor = colors.dialogBackground,
        title = { Text(title, color = colors.dialogText) },
        text = { Text(message, color = colors.dialogMuted) },
        confirmButton = {
            TextButton(
                onClick = { dismiss(true) },
                colors = ButtonDefaults.textButtonColors(contentColor = colors.danger),
            ) { Text(actionLabel) }
        },
        dismissButton = {
            TextButton(
                onClick = { dismiss(false) },
                colors = ButtonDefaults.textButtonColors(contentColor = colors.accent),
            ) { Text("Cancel") }
        },
    )
} ?: false

class PickOption<T>(
    val value: T,
    val label: String,
    val subtitle: String? = null,
    val icon: ImageVector? = null,
    val isAction: Boolean = false,
)

suspend fun <T> pickFromList(
    title: String,
    options: List<PickOption<T>>,
    current: T,
): PickOption<T>? = Dialogs.show { dismiss ->
    val colors = appColors
    AlertDialog(
        onDismissRequest = { dismiss(null) },
        containerColor = colors.dialogBackground,
        title = { Text(title, color = colors.dialogText) },
        text = {
            LazyColumn(Modifier.heightIn(max = 400.dp)) {
                items(options) { option ->
                    val selected = !option.isAction && option.value == current
                    val tint = if (option.isAction) colors.accent else colors.dialogText
                    ListItem(
                        headlineContent = {
                            Text(
                                option.label,
                                color = tint,
                                fontSize = 14.sp,
                                fontWeight = if (option.isAction) FontWeight.W600 else FontWeight.W400,
                            )
                        },
                        supportingContent = option.subtitle?.let {
                            { Text(it, color = colors.dialogMuted, fontSize = 12.sp) }
                        },
                        leadingContent = option.icon?.let {
                            {
                                Icon(
                                    it,
                                    contentDescription = null,
                                    tint = if (option.isAction) colors.accent else colors.textMuted,
                                    modifier = Modifier.size(20.dp),
                                )
                            }
                        },
                        trailingContent = if (selected) {
                            {
                                Icon(
                                    Icons.Default.Check,
                                    contentDescription = null,
                                    tint = colors.accent,
                                    modifier = Modifier.size(18.dp),
                                )
                            }
                        } else {
                            null
                        },
                        colors = ListItemDefaults.colors(containerColor = colors.dialogBackground),
                        modifier = Modifier.clickableRow { dismiss(option) },
                    )
                }
            }
        },
        confirmButton = {},
        dismissButton = {
            TextButton(
                onClick = { dismiss(null) },
                colors = ButtonDefaults.textButtonColors(contentColor = colors.accent),
            ) { Text("Cancel") }
        },
    )
}

private fun Modifier.clickableRow(onClick: () -> Unit): Modifier =
    this.clickable(onClick = onClick)

suspend fun promptForText(
    title: String,
    label: String,
    initial: String = "",
    obscure: Boolean = false,
    actionLabel: String = "Save",
): String? = Dialogs.show { dismiss ->
    val colors = appColors
    var value by rememberSaveable { mutableStateOf(initial) }
    AlertDialog(
        onDismissRequest = { dismiss(null) },
        containerColor = colors.dialogBackground,
        title = { Text(title, color = colors.dialogText) },
        text = {
            QTextField(
                value = value,
                onValueChange = { value = it },
                label = label,
                obscure = obscure,
                autofocus = true,
                onSubmitted = { dismiss(value.trim()) },
            )
        },
        confirmButton = {
            TextButton(
                onClick = { dismiss(value.trim()) },
                colors = ButtonDefaults.textButtonColors(contentColor = colors.accent),
            ) { Text(actionLabel) }
        },
        dismissButton = {
            TextButton(
                onClick = { dismiss(null) },
                colors = ButtonDefaults.textButtonColors(contentColor = colors.accent),
            ) { Text("Cancel") }
        },
    )
}
